src/backends/interflop-mca-int/interflop_mca_int.c

```diff
@@ -0,0 +1,681 @@
+  //  next, shift the MSB all the way to the LSB position
+  noise_msb = noise_msb >> (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE);
+  // noise sign mask is used to create the one's complement of the masked noise
+  noise_sign_mask = (uint64_t)DOUBLE_MASK_ONE + (noise_msb ^ 1);
+
+  // printf("noise: %lu\n", noise);
+  // printf("noise mask: %lu\n", noise_mask);
+  // printf("noise msb: %lu\n", noise_msb);
+  // printf("noise sign mask: %lu\n", noise_sign_mask);
+
+  // apply the mask to the noise, to only keep the noise at the correct
+  // magnitude
+  noise &= noise_mask;
+  // create the two's complement of the noise, if necessary
+  //  flip all bits of the noise, if necessary
+  noise ^= noise_sign_mask;
+  //  add the carry, if necessary
+  noise += noise_msb;
+
+  // printf("noise masked: %lu\n", noise);
+
+  // printf("input: %.18f\n", x_b64.f64);
+
+  // add the noise to the input
+  x_b64.u64 = x_b64.u64 + noise;
+
+  // printf("input noised: %.18f\n", x_b64.f64);
+
+  *x = x_b64.f64;
+}
+
+/* noise = rand * 2^(exp) */
+/* We can skip special cases since we never met them */
+/* Since we have exponent of double values, the result */
+/* is comprised between: */
+/* 1023+1023 = 2046 < QUAD_EXP_MAX (16383)  */
+/* -1022-53+-1022-53 = -2200 > QUAD_EXP_MIN (-16382) */
+static __float128 _noise_binary128(__float128 *x, const int exp,
+                                   rng_state_t *rng_state) {
+  uint32_t mask_shift_amount_high, mask_shift_amount_low;
+  uint64_t noise_mask_high, noise_mask_low, noise_msb, noise_sign_mask;
+  uint64_t noise_high, noise_low, carry;
+  uint64_t noise_low_msb, noise_low_msb_flip, noise_low_sign_carry;
+  binary128 x_b128 = {.f128 = *x};
+
+  // amount by which to shift, when creating the mask for the noise
+  mask_shift_amount_high = 1 + QUAD_EXP_SIZE + exp;
+  // create the mask for the noise high part
+  noise_mask_high = ((uint64_t)DOUBLE_MASK_ONE) >> mask_shift_amount_high;
+
+  // amount by which to shift, when creating the mask for the noise
+  mask_shift_amount_low =
+      exp - (1 + DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE - (1 + QUAD_EXP_SIZE));
+  // create the mask for the noise low part
+  noise_mask_low = ((uint64_t)DOUBLE_MASK_ONE) >> mask_shift_amount_low;
+
+  // generate a new random 64-bit integer
+  noise_low = _get_rand_uint64(rng_state, &global_tid_lock, &global_tid);
+  noise_high = _get_rand_uint64(rng_state, &global_tid_lock, &global_tid);
+
+  // extract the MSB of the noise
+  //  first extract the MSB
+  noise_msb =
+      noise_high & (((uint64_t)1) << (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE));
+  //  next, shift the MSB all the way to the LSB position
+  noise_msb = noise_msb >> (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE);
+  // noise sign mask is used to create the one's complement of the masked noise
+  noise_sign_mask = (uint64_t)DOUBLE_MASK_ONE + (noise_msb ^ 1);
+
+  // apply the mask to the noise, to only keep the noise at the correct
+  // magnitude
+  noise_low &= noise_mask_low;
```

Comment:
I believe this code can also be simplified as described for the double case; by removing the special treatment of the noise bit.

To avoid having to deal with the two uint64_t, I would use the native GCC __int128 type.